// netsuite-search/lib/netsuite-search.js
var _ = require('underscore');
var async = require('async');
var rest = require('restler');
var server=require('server.js');
console.log('server:'+server);

var CryptoJS = require("crypto-js");
var OAuth   = require('oauth-1.0a');


module.exports = function(settings) {
  if (typeof settings === 'undefined' || !settings.netsuite) {
    throw("You must provide settings containing netsuite credentials");
  }

  var netsuite = settings.netsuite;

  function run(rest_param, callback) {
	  
	  console.log('inside run:');
	  
	  var rest_url_selector=server.rec_type;
	  //var rest_url_selector='ProfileView';
	  
//var rest_url_selector1='PurchaseOrder';
console.log('rest_url_selector:'+rest_url_selector);

if(rest_url_selector=='ProfileView')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=172&deploy=1';
}

else if(rest_url_selector=='Updateall')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=265&deploy=1';
}

else if(rest_url_selector=='PurchaseOrder')
//if(rest_url_selector1=='PurchaseOrder')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=175&deploy=1';
}

else if(rest_url_selector=='BillPayment')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=248&deploy=1';
}

else if(rest_url_selector=='NewVendor')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=239&deploy=1';
}

else if(rest_url_selector=='ItemReceipt')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=241&deploy=1';
}

else if(rest_url_selector=='PackingList')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=177&deploy=1';
}

else if(rest_url_selector=='Bills')
{
	var restlet = 'https://rest.netsuite.com/app/site/hosting/restlet.nl?script=178&deploy=1';
}
	
	//var  json_data = "&param="+JSON.stringify(rest_param);
	var  json_data = JSON.stringify(rest_param);
	console.log('rest_param::'+rest_param);
	console.log('json_data::'+json_data);
	
	var oauth = OAuth(
		{
			consumer: {
					public: '857f3237925c76744c925d35c178d5fc0645a3a9a3316aa409ca24bfaf218174',
					secret: 'df9f92576ad7566f5152dd014a284ebbc52cb79049126e0287f0a0c1e0a04810'
                },
			signature_method: 'HMAC-SHA1'
		});

		var request_data = 
		{
			url: restlet,
			method: 'POST',
			data: {}
		};

		var token =
			{
				public: '1a20c4ba3c3464c9b2a0d46ba10f2bd26db450a1097cde3e167d85fb9edcf63e',
				secret: '2927d308eeeaf1aff4c8dca874d71d85dcbec51727ac40d75e6799d95bf0d7ed'
            };
		console.log('Before Setting header');
		var headerWithRealm = oauth.toHeader(oauth.authorize(request_data, token));
		//console.log("Signature headerWithRealm.Authorization == "+headerWithRealm.Authorization);
		headerWithRealm.Authorization += ', realm="TSTDRV1024523"';
		console.log('Between Setting header');
		var hd = headerWithRealm.Authorization;
		//console.log("oauth_nonce == "+headerWithRealm.Authorization);
		console.log('After Setting header');
		var restOptions = 
		{
			headers:{'Authorization':hd,'Content-Type':'application/json'},
			data:json_data
		};
	
    function onError(error) { console.log("error reported:"+error.message);
	console.log('inside err:');
    } 
	
    
    //var  json_data = "&param="+JSON.stringify(rest_param);
    //console.log('json_data::'+json_data);
    //columnsQuery = columns ? "&columns="+encodeURIComponent(JSON.stringify(columns)) : ''

    //var queryString = "&type="+type+filtersQuery+columnsQuery;
	//console.log('Request::'+restOptions);
	
	//var  json_data = "&param="+encodeURIComponent(JSON.stringify(rest_param));
	  console.log('Before RestCall:');
	rest.post(restlet, restOptions).on('success', function (result) {
      callback(null, result);
	  console.log('inside rest call:');
    }).on('error', onError).on('fail', function (err) { onError(err.error); });
	
	/*
	rest.get(restlet + json_data, restOptions).on('success', function (result) {
      callback(null, result);
    }).on('error', onError).on('fail', function (err) { onError(err.error); });
	*/
  }

  return {
    run: run
  };
};
